tf_helpers/lax.py

```diff
@@ -0,0 +1,463 @@
+
+  Args:
+    lhs: an array (the left-hand side matrix/vector to be multiplied)
+    rhs: an array (the right-hand side matrix/vector to be multiplied)
+    dimension_numbers: (Tuple[Tuple[Sequence[int], Sequence[int]],
+      Tuple[Sequence[int], Sequence[int]]]) – a tuple of tuples of the form
+      ((lhs_contracting_dims, rhs_contracting_dims), (lhs_batch_dims, rhs_batch_dims))
+
+  Returns:
+    An array that contains the result.
+  """
+  char_list = list(string.ascii_lowercase)
+  char_list = char_list[8:] + char_list[:8]
+  lhs_rank, rhs_rank = len(lhs.shape), len(rhs.shape)
+  lhs_rep = char_list[:lhs_rank]
+  rhs_rep = char_list[lhs_rank:lhs_rank+rhs_rank]
+  contraction, batch = dimension_numbers
+  lhs_contraction, rhs_contraction = contraction
+  if len(lhs_contraction) != len(rhs_contraction):
+    raise ValueError("The input matrices are required to have the same number "
+                     "of contraction dimensions, but got: lhs {}, rhs: {}".format(
+                     len(lhs_contraction), len(rhs_contraction)))
+  lhs_batch, rhs_batch = batch
+  if len(lhs_batch) != len(rhs_batch):
+    raise ValueError("The input matrices are required to have the same number "
+                     "of batch dimensions, but got: lhs {}, rhs: {}".format(
+                     len(lhs_batch), len(rhs_batch)))
+
+  if len(lhs_batch) == 0 and len(rhs_batch) == 0:
+    return np.asarray(
+      tf.tensordot(lhs, rhs, axes=(list(lhs_contraction), list(rhs_contraction))))
+
+  if (lhs_rank == rhs_rank == 3 and lhs_batch == (0,) and rhs_batch == (0,)
+      and lhs_contraction == (2,) and rhs_contraction == (1,)):
+    return np.asarray(tf.linalg.matmul(lhs, rhs))
+
+  for i in range(len(lhs_contraction)):
+    rhs_rep[rhs_contraction[i]] = lhs_rep[lhs_contraction[i]]
+  for i in range(len(lhs_batch)):
+    rhs_rep[rhs_batch[i]] = lhs_rep[lhs_batch[i]]
+
+  output_rep = _compose_output_rep(lhs_rep, rhs_rep, lhs_contraction,
+                                  rhs_contraction, lhs_batch, rhs_batch)
+  equation = ''.join(lhs_rep) + ',' + ''.join(rhs_rep) + "->" + output_rep
+  return np.asarray(tf.einsum(equation, lhs, rhs))
+
+
+def reduce_window(inputs, init_value, reducer, window_dimensions, strides,
+                  padding, base_dilation=None, window_dilation=None):
+  # Add an extra "batch" dimension and an extra "channel" dimension to pass the
+  # TensorFlow pool dimensionality checker.
+  inputs = np.expand_dims(inputs, axis=(0, inputs.ndim))
+  if reducer not in [np.max, np.add]:
+    raise TypeError("Only max pooling and average/sum pooling are supported.")
+
+  # Note that there is no need to send in the parameter data format since the
+  # input is already of default data format - "N...C". The adjustments of the
+  # input shape is already finished in apply_fun of Pooling in stax.
+  pooling = "AVG" if reducer == np.add else "MAX"
+  output = nn.pool(inputs, window_dimensions, pooling, strides, padding)
```

Comment:
I think you also need to remove (see e.g. https://numpy.org/doc/stable/reference/generated/numpy.squeeze.html) the leading and training dimensions when you return the array too (so that you return the array of rank `N+2` and not `N+4`.

Reply:
My bad. I am really sorry about this.